### PR TITLE
Add the ability to pass in an already generated discharge token.

### PR DESCRIPTION
**Pull request by fabricematrat**
Add the ability to pass in an already generated discharge token.

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5919/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5920/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5921/
Test PASSed.

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -1578,12 +1579,13 @@ YUI.add('juju-gui', function(Y) {
       if (this.get('charmstore') === undefined) {
         var jujuConfig = window.juju_config;
         var charmstoreURL = '';
-        var existingMacaroons, existingCookie;
+        var existingMacaroons, existingCookie, existingDischargeToken;
```
Comment (frankban) [inline jujugui/static/gui/src/app/app.js]: While you are here, s/var/let/

File: jujugui/static/gui/src/app/store/env/bakery.js [hunk 2]
```diff
@@ -571,8 +571,11 @@ YUI.add('juju-env-bakery', function(Y) {
       _exportMacaroon: function (successCallback, failureCallback, response) {
         try {
           var json = JSON.parse(response.target.responseText);
-          this.dischargeStore.setItem(DISCHARGE_TOKEN,
-            btoa(JSON.stringify(json.DischargeToken)));
+          if (json.DischargeToken !== undefined &&
```
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/bakery.js]: just `if (json.DischargeToken)` would work?

File: jujugui/views.py [hunk 3]
```diff
@@ -74,6 +74,7 @@ def config(request):
     env_uuid = settings.get('jujugui.uuid', 'sandbox')
 
     charmstore_macaroons = settings.get('jujugui.charmstore_macaroons')
+    discharge_token = settings.get('jujugui.discharge_token')
```
Comment (frankban) [inline jujugui/views.py]: Please add a test for this new option.

Comment (fabricematrat) [inline jujugui/views.py]: done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5923/
Test PASSed.

File: jujugui/views.py [hunk 4]
```diff
@@ -74,6 +74,7 @@ def config(request):
     env_uuid = settings.get('jujugui.uuid', 'sandbox')
 
     charmstore_macaroons = settings.get('jujugui.charmstore_macaroons')
```
Comment (jaycee) [inline jujugui/views.py]: Do we still need the charmstore_macaroon to be passed via settings, or with the idm token can we handle all the login stuff like we do in HJC?

File: jujugui/views.py [hunk 5]
```diff
@@ -74,6 +74,7 @@ def config(request):
...
+    discharge_token = settings.get('jujugui.discharge_token')
     plans_macaroons = settings.get('jujugui.plans_macaroons')
     terms_macaroons = settings.get('jujugui.terms_macaroons')
```
Comment (jaycee) [inline jujugui/views.py]: Same question about passing in macaroons for plans and terms...if we're doing non-interactive login to idm with the token, can't we let that handle all bakery requests absent passed in macaroons?

File: jujugui/views.py (hunk 4 above)
Comment (fabricematrat) [inline jujugui/views.py]: when we switch to gui handling all the authenticated data.

**Comment by fabricematrat** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**